File: src/ai_py/AIRouter.py
from __future__ import annotations

import os
import time
from typing import Optional

from .providers import ProviderCerebras, ProviderGroq, ProviderOpenRouter
from .types import AIContext, AIProvider, AIProviderError, AI_FRIENDLY_ERROR


class AIRouter:
    _providers: dict[str, AIProvider] = {
        "Groq": ProviderGroq(),
        "Cerebras": ProviderCerebras(),
        "OpenRouter": ProviderOpenRouter(),
    }
    _backoff_ms = [300, 600, 1000]

    @classmethod
    def handle(cls, input_text: str, context: Optional[AIContext] = None) -> str:
        last_error: Optional[Exception] = None
        provider_order = cls._provider_order()

        for index, provider in enumerate(provider_order):
            print(f"[AI Service] Iniciando requisição com provedor: {provider.name}")
            try:
                return cls._run_with_retries(provider, input_text, context)
            except Exception as exc:
                last_error = exc
                next_provider = provider_order[index + 1] if index + 1 < len(provider_order) else None
                if next_provider:
                    print(
                        f"[AI Critical] {provider.name} esgotou tentativas. "
                        f"Chaveando dinamicamente para Fallback {index + 1}: {next_provider.name}"
                    )

        print(f"[AI Critical] Todos os provedores falharam. Ultimo erro: {last_error}")
        return AI_FRIENDLY_ERROR

    @classmethod
    def _run_with_retries(cls, provider: AIProvider, input_text: str, context: Optional[AIContext]) -> str:
        max_attempts = min(max(cls._int_env("AI_PROVIDER_MAX_ATTEMPTS", 3), 2), 3)
        for attempt in range(1, max_attempts + 1):
            try:
                return provider.generate_response(input_text, context)
            except AIProviderError as exc:
                is_last = attempt == max_attempts
                if not exc.retryable or is_last:
                    raise
                status_text = f"código {exc.status_code}" if exc.status_code else str(exc)
                print(f"[AI Warning] {provider.name} falhou com {status_text}. Tentando novamente (Retry #{attempt})...")
                time.sleep(cls._backoff_ms[min(attempt - 1, len(cls._backoff_ms) - 1)] / 1000)

        raise AIProviderError(f"{provider.name} esgotou tentativas.", provider.name, retryable=True)

    @classmethod
    def _provider_order(cls) -> list[AIProvider]:
        default_order = ["Groq", "Cerebras", "OpenRouter"]
        env_order = [
            os.getenv("AI_PROVIDER_PRIMARY"),
            os.getenv("AI_PROVIDER_FALLBACK_1"),
            os.getenv("AI_PROVIDER_FALLBACK_2"),
        ]
        names = []
        for index, name in enumerate(env_order):
            normalized = cls._normalize_provider_name(name) or default_order[index]
            if normalized not in names:
                names.append(normalized)
        return [cls._providers[name] for name in names if name in cls._providers]

    @staticmethod
    def _normalize_provider_name(name: Optional[str]) -> Optional[str]:
        value = (name or "").strip().lower()
        if value == "groq":
            return "Groq"
        if value == "cerebras":
            return "Cerebras"
        if value in ("openrouter", "open-router"):
            return "OpenRouter"
        return None

    @staticmethod
    def _int_env(key: str, fallback: int) -> int:
        try:
            value = int(os.getenv(key, ""))
            return value if value > 0 else fallback
        except Exception:
            return fallback

File: src/ai/AIRouter.ts
import { ProviderGroq } from "./providers/ProviderGroq";
import { ProviderCerebras } from "./providers/ProviderCerebras";
import { ProviderOpenRouter } from "./providers/ProviderOpenRouter";
import {
  AIContext,
  AIProvider,
  AIProviderError,
  AIProviderName,
  readOptionalNumberEnv,
} from "./types";

const FRIENDLY_AI_ERROR = "Serviço de IA temporariamente indisponível.";
const BACKOFF_MS = [300, 600, 1000];

export class AIRouter {
  private static readonly providers: Record<AIProviderName, AIProvider> = {
    Groq: new ProviderGroq(),
    Cerebras: new ProviderCerebras(),
    OpenRouter: new ProviderOpenRouter(),
  };

  static async handle(input: string, context?: AIContext): Promise<string> {
    const providerOrder = this.resolveProviderOrder();
    let lastError: unknown = null;

    for (const [index, provider] of providerOrder.entries()) {
      console.log(`[AI Service] Iniciando requisição com provedor: ${provider.name}`);

      try {
        return await this.runWithRetries(provider, input, context);
      } catch (error) {
        lastError = error;
        const nextProvider = providerOrder[index + 1];

        if (nextProvider) {
          console.error(
            `[AI Critical] ${provider.name} esgotou tentativas. Chaveando dinamicamente para Fallback ${index + 1}: ${nextProvider.name}`,
          );
        }
      }
    }

    console.error("[AI Critical] Todos os provedores falharam.", lastError);
    return FRIENDLY_AI_ERROR;
  }

  private static async runWithRetries(
    provider: AIProvider,
    input: string,
    context?: AIContext,
  ): Promise<string> {
    const maxAttempts = Math.min(Math.max(readOptionalNumberEnv("AI_PROVIDER_MAX_ATTEMPTS", 3), 2), 3);

    for (let attempt = 1; attempt <= maxAttempts; attempt += 1) {
      try {
        return await provider.generateResponse(input, context);
      } catch (error) {
        const providerError = this.normalizeError(error, provider.name);
        const isLastAttempt = attempt === maxAttempts;

        if (!providerError.retryable || isLastAttempt) {
          throw providerError;
        }

        const statusText = providerError.statusCode ? `código ${providerError.statusCode}` : providerError.message;
        console.warn(
          `[AI Warning] ${provider.name} falhou com ${statusText}. Tentando novamente (Retry #${attempt})...`,
        );

        await this.sleep(BACKOFF_MS[attempt - 1] ?? BACKOFF_MS[BACKOFF_MS.length - 1]);
      }
    }

    throw new AIProviderError(`${provider.name} esgotou tentativas.`, provider.name, true);
  }

  private static normalizeError(error: unknown, providerName: string): AIProviderError {
    if (error instanceof AIProviderError) {
      return error;
    }

    return new AIProviderError(
      error instanceof Error ? error.message : "Erro desconhecido no provedor.",
      providerName,
      true,
    );
  }

  private static resolveProviderOrder(): AIProvider[] {
    const envOrder = [
      process.env.AI_PROVIDER_PRIMARY,
      process.env.AI_PROVIDER_FALLBACK_1,
      process.env.AI_PROVIDER_FALLBACK_2,
    ];
    const defaultOrder: AIProviderName[] = ["Groq", "Cerebras", "OpenRouter"];
    const selectedNames = envOrder
      .map((name, index) => this.normalizeProviderName(name) ?? defaultOrder[index])
      .filter((name): name is AIProviderName => Boolean(name));

    const uniqueNames = [...new Set(selectedNames)];
    return uniqueNames.map((name) => this.providers[name]);
  }

  private static normalizeProviderName(name?: string): AIProviderName | null {
    const normalized = (name || "").trim().toLowerCase();
    if (!normalized) return null;
    if (normalized === "groq") return "Groq";
    if (normalized === "cerebras") return "Cerebras";
    if (normalized === "openrouter" || normalized === "open-router") return "OpenRouter";
    return null;
  }

  private static sleep(ms: number): Promise<void> {
    return new Promise((resolve) => setTimeout(resolve, ms));
  }
}
